### PR TITLE
Xcom gcs ds

### PR DESCRIPTION
**Pull request by fhoda**
- Updated the dag to have a more natural looking flow and access XCom accordingly for the `fit` task.
- Added more comments as well

### CODE REVIEW

File: dags/xcom_gcs_ds.py [hunk 1]
```diff
@@ -33,6 +23,8 @@ def using_gcs_for_xcom_ds():
 
     @task
     def load_data():
+        """Pull Census data from Public BigQuery and save as Pandas dataframe in GCS bucket with XCom"""
+
         client = bigquery.Client()
```
Comment (virajmparekh) [inline dags/xcom_gcs_ds.py]: https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/hooks/bigquery.py#L204

Can we use the hook here with the get df method? It'll simplify the `requirements.txt` as well to only need the Google provider

Comment (fhoda) [reply dags/xcom_gcs_ds.py]: Ohhh...did not know that function existed! Yes, will update.

File: dags/xcom_gcs_ds.py [hunk 2]
```diff
@@ -105,8 +123,22 @@ def train(df: pd.DataFrame):
...
+        Returns lightgbm model as json via XCom to GCS bucket.
+        
+
+        Keyword arguments:
+        accuracy -- average accuracy score as determined by CV. 
+        """
         if accuracy >= .8:
+
+            # Reuse data produced by the feauture_engineering task by pulling from GCS bucket via XCom
+            df = kwargs['ti'].xcom_pull(task_ids='feature_engineering')
```
Comment (santonaastronomer) [inline dags/xcom_gcs_ds.py]: It bothers me a little bit that we're both using the taskflow API as well as doing a xcom_pull. I mean, I don't think the usage is wrong, it just takes away the 'magic' a little.

**Review by santonaastronomer — APPROVED** [review]

File: dags/xcom_gcs_ds.py (hunk 2 above)
Comment (virajmparekh) [reply dags/xcom_gcs_ds.py]: Hm - yeah good catch. @fhoda is there a more natural way to pull this?

Comment (fhoda) [reply dags/xcom_gcs_ds.py]: This is the only way I get the graph look like what we discussed while also reusing data from a task not directly prior to this one. The previous way I had it was more natural to Taskflow and would address the concern here but then take us back to where we started in terms of the rendering of the graph. This is my first use of Taskflow so will continue to experiment.

**Review by virajmparekh — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -6,8 +6,7 @@ In an effort to keep the examples easy to follow, much of the data processing an
...
-1. `xcom_gcs_ds.py` - DS/ML pipeline from data extraction to modeling with Python Operators and GCS Xcom backend.
-    - Pulls data from Google BigQuery into a pandas dataframe, prepares data, train and then builds model. 
-    - All intermediate data after each task is saved to GCS using Xcom and is also passed between tasks using this method.
-    - Model output is saved to GCS via Xcom.
-    - All tasks use the python operator.
\ No newline at end of file
+1. `xcom_gcs_ds.py` - A simple DS pipeline from data extraction to modeling.
+    - Pulls data from BigQuery using the Google Provider (BigQueryHook) into a dataframe that preps, trains, and builds the model
+    - Data is passed between the tasks using [XComs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html)
+    - Uses GCS as an XCOM backend to easily track intermediary data in a scalable, external system
```
Comment (josh-fell) [inline README.md]: Can "XCOM" be changed to "XCom"?

File: dags/xcom_gcs_ds.py [hunk 4]
```diff
@@ -3,25 +3,19 @@
...
 from sklearn.model_selection import cross_val_score
-from sklearn.metrics import classification_report, confusion_matrix
 
 from lightgbm import LGBMClassifier
 
-from google.cloud import bigquery
+from airflow.providers.google.cloud.hooks.bigquery import BigQueryHook
 
+docs = """
+By default, Airflow stores all return values in XCom. However, this can introduce complexity, as users then have to consider the size of data they are returning. Futhermore, since xcoms are stored in the Airflow database by default, intermediary data is not easily accessible by external systems.
```
Comment (josh-fell) [inline dags/xcom_gcs_ds.py]: Same comment for XCom spelling -> "xcoms" to "XComs".

File: dags/xcom_gcs_ds.py [hunk 5]
```diff
@@ -3,25 +3,19 @@
 from airflow.utils.dates import days_ago
 
 from datetime import datetime
-import json
```
Comment (josh-fell) [inline dags/xcom_gcs_ds.py:6]: The `airflow.models import DAG` statement can be removed since the `@dag` decorator is being used.

File: dags/xcom_gcs_ds.py [hunk 6]
```diff
@@ -3,25 +3,19 @@
...
-from google.cloud import bigquery
+from airflow.providers.google.cloud.hooks.bigquery import BigQueryHook
 
+docs = """
+By default, Airflow stores all return values in XCom. However, this can introduce complexity, as users then have to consider the size of data they are returning. Futhermore, since xcoms are stored in the Airflow database by default, intermediary data is not easily accessible by external systems.
+By using an external XCom backend, users can easily push and pull all intermediary data generated in their DAG in GCS.
+"""
 
 @dag(
     default_args={'owner': 'airflow'},
```
Comment (josh-fell) [inline dags/xcom_gcs_ds.py]: FYI - There is an ongoing effort to clean up how `default_args` are used in example DAGs for OSS but here, this value is the default. This can be removed.

Let's make the `start_date` value a static date rather than a dynamic one.  This is being cleaned up in example DAGs as well.

File: dags/xcom_gcs_ds.py (hunk 2 above)
Comment (josh-fell) [reply dags/xcom_gcs_ds.py]: I assume you all don't want the extra edge between "feature_engineering" and "fit" when passing the output of "feature_engineering" into this function?
image

Understandable but I agree it's not taking full advantage of the TaskFlow API.  However, you can update the function signature to directly access the `TaskInstance` object without changing how it's called:
```python
def fit(accuracy: float, ti=None): 
    ...

    # Reuse data produced by the feauture_engineering task by pulling from GCS bucket via XCom
    df = ti.xcom_pull(task_ids='feature_engineering')
    ...
```

File: dags/xcom_gcs_ds.py [hunk 7]
```diff
@@ -118,12 +153,17 @@ def fit(accuracy: float, df: pd.date_range):
...
-dag = using_gcs_for_xcom_ds()
\ No newline at end of file
+    accuracy = train(features)
+    fit(accuracy)
+
+    # Alternate method to set up task dependencies
+    # fit(train(feature_engineering(preprocessing(load_data()))))
+    
+dag = using_gcs_for_xcom_ds()
+dag.doc_md = docs
```
Comment (josh-fell) [inline dags/xcom_gcs_ds.py]: The `doc_md` arg can be passed directly into the `@dag` decorator:
```python
@dag(
    ...
    schedule_interval=None,
    catchup=False,
    doc_md=docs,
)
```

File: dags/xcom_gcs_ds.py (hunk 2 above)
Comment (josh-fell) [reply dags/xcom_gcs_ds.py]: This also _may_ be an option but I'm not sure how this will work when the desired input would be a `Dataframe`. If this approach doesn't work, I'd be skeptical if setting `render_template_as_native_obj` to True at the DAG level would properly render the value as a `Dataframe` since it may not be considered a "native Python type".

```python
@task
    def fit(accuracy: float, feature_dataset: pd.DataFrame, **kwargs):
...
fit(accuracy, feature_dataset="{{ ti.xcom_pull(task_ids='feature_engineering') }}")
```

Comment (santonaastronomer) [reply dags/xcom_gcs_ds.py]: So... since xcom is pulling the data from the same location, couldn't we make both data passages (train->fit and feature_engineering->fit) equally native? What I am suggesting is a change to the taskflow api rather than one to this PR. It should be relatively straightforward to abstract away `ti.xcom_pull(task_ids='feature_engineering')` by updating the `@task` decorator. Eventually we should support pulling featurized data freely from a feature store even if it's not part of the same branch of the DAG (or even the same DAG). WDYT @josh-fell?

Comment (josh-fell) [reply dags/xcom_gcs_ds.py]: 100% agree. Even the abstraction included with the `.output` property of operators and `XComArgs` doesn't exactly mirror the classic `xcom_pull()` functionality which gives users access to `XComs` across DAG and execution dates plus access to specific items directly in an `XCom`.  An abstraction you are putting forth would be highly beneficial and I'd love to see it implemented.  Perhaps Daniel has something like this baking with his current decorator-wizardry activities.

I was only suggesting some alternatives if we didn't want to show a non-serial pipeline path which is generated when we try to `fit(accuracy, features)`.

Comment (fhoda) [reply dags/xcom_gcs_ds.py]: I have applied @josh-fell 's first suggestion. I am not really opinionated on this matter and can test out the second one if there are strong feelings towards using that syntax.

**Review by josh-fell — APPROVED** [review]
👍
